src/inmanta/server/services/orchestrationservice.py

```diff
@@ -419,6 +427,151 @@ def safe_get(input: JsonType, key: str, default: object) -> object:
+        :param resource_sets: a dictionary mapping each resource (using its id) with a resource_set if it is in one
+        :param removed_resource_sets: The names of the resource sets removed in this partial compile.
+        """
+
+        def pair_resources_partial_update_to_old_version(
+            old_resources: List[tuple[Dict[str, Any], Optional[str]]], partial_updates: List[Dict[str, Any]]
+        ) -> List[tuple[Dict[str, Any], Optional[Dict[str, Any]], Optional[str], Optional[str]]]:
+            """
+            returns a list of tuples (one tuple per resource) containing 4 elements:
+            1. the resource as returned by the partial compile
+            2. the resource as it is in the old version
+            3. the resource_set the resource belongs to in the partial compile
+            4. the resource_set the resource belongs to in the previous version
+
+            :param old_resources: a list of tuples containing the resource in the previous version of the model
+            and its resource_set.
+            :param partial_updates: The list of resources part of the partial compile.
+            """
+            paired_resources: List[tuple[Dict[str, Any], Optional[Dict[str, Any]], Optional[str], Optional[str]]] = []
+            for partial_update in partial_updates:
+                key = Id.parse_id(partial_update["id"]).resource_str()
+                resource_set = resource_sets.get(key)
+                pair: tuple[Dict[str, Any], Optional[Dict[str, Any]], Optional[str], Optional[str]] = (
+                    partial_update,
+                    None,
+                    resource_set,
+                    None,
+                )
+                for old_resource in old_resources:
+                    res = old_resource[0]
+                    if key == Id.parse_id(res["id"]).resource_str():
+                        pair = (partial_update, res, resource_set, old_resource[1])
+                paired_resources.append(pair)
+            return paired_resources
+
+        def get_updated_resource_sets(
+            paired_resources: List[tuple[Dict[str, Any], Optional[Dict[str, Any]], Optional[str], Optional[str]]]
+        ) -> Set[str]:
+            """
+            returns a set of strings containing the name of all the resource_sets that have changed resources
+            in the result of the partial compile.
+
+            :param paired_resources: see pair_resources_partial_update_to_old_version
+            """
+            result: Set[str] = set()
+            for res in paired_resources:
+                resource_set = res[2]
+                if resource_set:
+                    result.add(resource_set)
+            return result
+
+        async def get_old_resources() -> List[tuple[Dict[str, Any], Optional[str]]]:
+            old_data = await data.Resource.get_list()
+            result: List[tuple[Dict[str, Any], Optional[str]]] = []
```

Comment:
This is quite a complicated type definition. The problem with this is that it doesn't really explain what is inside the data structure. In other words, the code becomes hard to read. It would be good to introduce a new class (or a dataclass) here.